functorch/experimental/map.py

```diff
@@ -0,0 +1,97 @@
+import torch
+import torch.utils._pytree as pytree
+from torch._C import DispatchKey, DispatchKeySet, ExcludeDispatchKeyGuard
+from torch._ops import PyOperator
+from torch._subclasses.fake_tensor import FakeTensorMode
+from torch.fx.experimental.proxy_tensor import (
+    disable_proxy_modes_tracing,
+    get_proxy_slot,
+    make_fx,
+    ProxyTorchDispatchMode,
+    track_tensor_tree,
+)
+from torch.utils._python_dispatch import (
+    _get_current_dispatch_mode,
+    _pop_mode_temporarily,
+)
+from torch.utils._pytree import tree_flatten
+
+
+map = PyOperator("map")
+
+
+def trace_map(proxy_mode, func_overload, f, xs, *args):
+    def _unwrap_proxy(e):
+        if not isinstance(e, (torch.Tensor, torch.SymInt, torch.SymFloat)):
+            return e
+        return get_proxy_slot(e, proxy_mode.tracer, e, lambda e: e.proxy)
+
+    assert isinstance(xs, torch.Tensor), "map() must loop over a tensor"
+    assert len(xs.shape) > 0 and xs.shape[0] > 0, "map() cannot be traced with scalar tensors or zero dimension tensors"
+    assert all(isinstance(o, (torch.Tensor, torch.nn.Module)) for o in args), "map() operands must be a list of tensors of modules"
+
+    with disable_proxy_modes_tracing():
+        body_graph = make_fx(f)(xs[0], *args)
+
+    next_name = None
+    i = 0
+    while not next_name:
+        candidate = f"body_graph_{i}"
+        if hasattr(proxy_mode.tracer.root, candidate):
+            i += 1
+        else:
+            next_name = candidate
+
+    proxy_mode.tracer.root.register_module(next_name, body_graph)
+    node_args = (body_graph, xs, *args)
+    proxy_args = pytree.tree_map(_unwrap_proxy, node_args)
+    out_proxy = proxy_mode.tracer.create_proxy('call_function', func_overload, proxy_args, {},
+                                               name="map")
+    outs = [body_graph(x, *args) for x in xs]
+    out = xs.new_empty([xs.shape[0], *outs[0].shape])
+    out.copy_(torch.stack(outs))
```

Comment:
why do we need to construct `out` this way? Why couldn't it have been `out = torch.stack(outs)`?

Reply:
@zou3519 this is good question and the reason I didn't use `torch.stack()` is because `torch.stack([...])` will take a Python list which doesn't have the dynamic shape information we want to preserve. For example:
```
xs: Tensor[s0, s1]
l = [x for x in xs]
ys = torch.stack(l)
# ys will be specialized to: Tensor[c, s1] where c is the 
```
ideally we want `ys` to preserve dynamic shape dimensions, e.g. `Tensor[s0, s1]` here as well.

maybe there could be a better way but currently I only got `new_empty()` + `copy_` to work.

Reply:
I see the problem...

Reply:
Okay, this seems reasonable to me. It's not like we're actually executing new_empty and then copy_ (which would do one additional tensor allocation that is not necessary), and at this point the only thing we have Tensors (and not Tensor subclasses, where the copy_ would have been problematic).

Could we add a comment in the code about this? (otherwise, someone might do the rewrite to torch.stack and break something)

Reply:
ok np